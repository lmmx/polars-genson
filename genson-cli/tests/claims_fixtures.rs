use assert_cmd::Command;
use insta::{assert_snapshot, with_settings};
use std::fs;

/// Check if the current output matches the verified/blessed version
fn is_output_approved(snapshot_name: &str, output: &str) -> bool {
    let module_file = file!();
    let module_stem = std::path::Path::new(module_file)
        .file_stem()
        .unwrap()
        .to_string_lossy();
    let verified_path = format!("tests/verified/{}__{}.snap", module_stem, snapshot_name);

    if let Ok(verified_content) = fs::read_to_string(&verified_path) {
        if let Some(header_end) = verified_content.find("\n---\n") {
            let verified_output = &verified_content[header_end + 5..];
            return verified_output.trim() == output.trim();
        }
    }
    false
}

/// Run genson-cli with claims fixture from disk
fn run_genson_claims_fixture_from_disk(fixture_path: &str, name: &str, extra_args: &[&str]) {
    let mut cmd = Command::cargo_bin("genson-cli").unwrap();
    let mut args = vec![
        "--map-threshold",
        "0",
        "--unify-maps",
        "--wrap-root",
        "claims",
    ];
    // Add --ndjson flag if the fixture is a .jsonl file
    if fixture_path.ends_with(".jsonl") {
        args.push("--ndjson");
    }
    args.extend_from_slice(extra_args);
    args.push(fixture_path);
    let args_for_metadata = args.clone();
    cmd.args(args);

    let assert_output = cmd.assert().success();
    let output = assert_output.get_output();
    let stdout_str = String::from_utf8(output.stdout.clone()).unwrap();

    // Let stderr be visible if there's any debug output
    if !output.stderr.is_empty() {
        let stderr_str = String::from_utf8_lossy(&output.stderr);
        eprintln!("stderr from {}: {}", name, stderr_str);
    }

    let approved = is_output_approved(name, &stdout_str);

    with_settings!({
        info => &serde_json::json!({
            "approved": approved,
            "args": args_for_metadata[..args_for_metadata.len()-1], // exclude file path
            "fixture": fixture_path
        })
    }, {
        assert_snapshot!(name, stdout_str);
    });
}

#[test]
fn test_claims_fixture_l1_avro() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L1.jsonl",
        "claims_fixture_l1__avro",
        &["--avro"],
    );
}

#[test]
fn test_claims_fixture_l1_jsonschema() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L1.jsonl",
        "claims_fixture_l1__jsonschema",
        &[],
    );
}

#[test]
fn test_claims_fixture_l1_normalize() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L1.jsonl",
        "claims_fixture_l1__normalize",
        &["--normalise"],
    );
}

#[test]
fn test_claims_fixture_l2_avro() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L2.jsonl",
        "claims_fixture_l2__avro",
        &["--avro"],
    );
}

#[test]
fn test_claims_fixture_l2_jsonschema() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L2.jsonl",
        "claims_fixture_l2__jsonschema",
        &[],
    );
}

#[test]
fn test_claims_fixture_l2_normalize() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L2.jsonl",
        "claims_fixture_l2__normalize",
        &["--normalise"],
    );
}

#[test]
fn test_claims_fixture_l3_avro() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L3.jsonl",
        "claims_fixture_l3__avro",
        &["--avro"],
    );
}

#[test]
fn test_claims_fixture_l3_jsonschema() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L3.jsonl",
        "claims_fixture_l3__jsonschema",
        &[],
    );
}

#[test]
fn test_claims_fixture_l3_normalize() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L3.jsonl",
        "claims_fixture_l3__normalize",
        &["--normalise"],
    );
}

#[test]
fn test_claims_fixture_l4_avro() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L4.jsonl",
        "claims_fixture_l4__avro",
        &["--avro"],
    );
}

#[test]
fn test_claims_fixture_l4_jsonschema() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L4.jsonl",
        "claims_fixture_l4__jsonschema",
        &[],
    );
}

#[test]
fn test_claims_fixture_l4_normalize() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L4.jsonl",
        "claims_fixture_l4__normalize",
        &["--normalise"],
    );
}

// Following are numbered from the full x1818 fixture

#[test]
fn test_claims_fixture_x1818_l4_avro() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L4_MINIMAL.json",
        "claims_fixture_x1818_l4__avro",
        &["--avro"],
    );
}

#[test]
fn test_claims_fixture_x1818_l4_jsonschema() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L4_MINIMAL.json",
        "claims_fixture_x1818_l4__jsonschema",
        &[],
    );
}

#[test]
fn test_claims_fixture_x1818_l4_normalize() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L4_MINIMAL.json",
        "claims_fixture_x1818_l4__normalize",
        &["--normalise"],
    );
}

#[test]
fn test_claims_fixture_x1818_l5_avro() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L5_MINIMAL.json",
        "claims_fixture_x1818_l5__avro",
        &["--avro"],
    );
}

#[test]
fn test_claims_fixture_x1818_l5_jsonschema() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L5_MINIMAL.json",
        "claims_fixture_x1818_l5__jsonschema",
        &[],
    );
}

#[test]
fn test_claims_fixture_x1818_l5_normalize() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L5_MINIMAL.json",
        "claims_fixture_x1818_l5__normalize",
        &["--normalise"],
    );
}

#[test]
fn test_claims_fixture_x1818_l12_avro() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L12_MINIMAL.json",
        "claims_fixture_x1818_l12__avro",
        &["--avro"],
    );
}

#[test]
fn test_claims_fixture_x1818_l12_jsonschema() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L12_MINIMAL.json",
        "claims_fixture_x1818_l12__jsonschema",
        &[],
    );
}

#[test]
fn test_claims_fixture_x1818_l12_normalize() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L12_MINIMAL.json",
        "claims_fixture_x1818_l12__normalize",
        &["--normalise"],
    );
}

#[test]
fn test_claims_fixture_x1818_l14_avro() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L14_MINIMAL.json",
        "claims_fixture_x1818_l14__avro",
        &["--avro"],
    );
}

#[test]
fn test_claims_fixture_x1818_l14_jsonschema() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L14_MINIMAL.json",
        "claims_fixture_x1818_l14__jsonschema",
        &[],
    );
}

#[test]
fn test_claims_fixture_x1818_l14_normalize() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L14_MINIMAL.json",
        "claims_fixture_x1818_l14__normalize",
        &["--normalise"],
    );
}

#[test]
fn test_claims_fixture_x1818_l16_avro() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L16_MINIMAL.json",
        "claims_fixture_x1818_l16__avro",
        &["--avro"],
    );
}

#[test]
fn test_claims_fixture_x1818_l16_jsonschema() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L16_MINIMAL.json",
        "claims_fixture_x1818_l16__jsonschema",
        &[],
    );
}

#[test]
fn test_claims_fixture_x1818_l16_normalize() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L16_MINIMAL.json",
        "claims_fixture_x1818_l16__normalize",
        &["--normalise"],
    );
}

#[test]
fn test_claims_fixture_x1818_l26_avro() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L26_MINIMAL.json",
        "claims_fixture_x1818_l26__avro",
        &["--avro"],
    );
}

#[test]
fn test_claims_fixture_x1818_l26_jsonschema() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L26_MINIMAL.json",
        "claims_fixture_x1818_l26__jsonschema",
        &[],
    );
}

#[test]
fn test_claims_fixture_x1818_l26_normalize() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L26_MINIMAL.json",
        "claims_fixture_x1818_l26__normalize",
        &["--normalise"],
    );
}
