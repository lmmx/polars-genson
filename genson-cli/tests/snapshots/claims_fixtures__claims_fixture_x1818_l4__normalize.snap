---
source: genson-cli/tests/claims_fixtures.rs
expression: stdout_str
info:
  approved: false
  args:
    - "--map-threshold"
    - "0"
    - "--unify-maps"
    - "--wrap-root"
    - claims
    - "--normalise"
  fixture: tests/data/claims/x1818_L4_MINIMAL.json
---
[
  {
    "claims": {
      "P119": [
        {
          "mainsnak": {
            "datavalue": {
              "id": "Q533697",
              "labels": {
                "it": "cimitero di Highgate"
              },
              "datavalue__string": null
            }
          },
          "qualifiers": {
            "P625": [
              {
                "datavalue": {
                  "precision": {
                    "precision__integer": null,
                    "precision__number": 0.00002777777777777778
                  },
                  "time": null,
                  "timezone": null,
                  "before": null,
                  "after": null,
                  "calendarmodel": null,
                  "latitude": 51.56652777777778,
                  "longitude": -0.14544444444444443,
                  "altitude": null,
                  "globe": "http://www.wikidata.org/entity/Q2"
                }
              }
            ]
          },
          "references": [
            {
              "P854": [
                {
                  "datavalue": {
                    "id": null,
                    "labels": null,
                    "datavalue__string": "http://highgatecemetery.org/visit/who",
                    "time": null,
                    "timezone": null,
                    "before": null,
                    "after": null,
                    "precision": null,
                    "calendarmodel": null,
                    "text": null,
                    "language": null
                  }
                }
              ],
              "P407": [
                {
                  "datavalue": {
                    "id": "Q1860",
                    "labels": {
                      "tg": "ингилисӣ"
                    },
                    "datavalue__string": null,
                    "time": null,
                    "timezone": null,
                    "before": null,
                    "after": null,
                    "precision": null,
                    "calendarmodel": null,
                    "text": null,
                    "language": null
                  }
                }
              ],
              "P813": [
                {
                  "datavalue": {
                    "id": null,
                    "labels": null,
                    "datavalue__string": null,
                    "time": "+2013-12-07T00:00:00Z",
                    "timezone": 0,
                    "before": 0,
                    "after": 0,
                    "precision": 11,
                    "calendarmodel": "http://www.wikidata.org/entity/Q1985727",
                    "text": null,
                    "language": null
                  }
                }
              ],
              "P1476": [
                {
                  "datavalue": {
                    "id": null,
                    "labels": null,
                    "datavalue__string": null,
                    "time": null,
                    "timezone": null,
                    "before": null,
                    "after": null,
                    "precision": null,
                    "calendarmodel": null,
                    "text": "Who’s here",
                    "language": "en"
                  }
                }
              ],
              "P248": [
                {
                  "datavalue": {
                    "id": "Q533697",
                    "labels": {
                      "it": "cimitero di Highgate"
                    },
                    "datavalue__string": null,
                    "time": null,
                    "timezone": null,
                    "before": null,
                    "after": null,
                    "precision": null,
                    "calendarmodel": null,
                    "text": null,
                    "language": null
                  }
                }
              ],
              "P1065": [
                {
                  "datavalue": {
                    "id": null,
                    "labels": null,
                    "datavalue__string": "https://web.archive.org/web/20131203025040/http://highgatecemetery.org/visit/who",
                    "time": null,
                    "timezone": null,
                    "before": null,
                    "after": null,
                    "precision": null,
                    "calendarmodel": null,
                    "text": null,
                    "language": null
                  }
                }
              ],
              "P2960": [
                {
                  "datavalue": {
                    "id": null,
                    "labels": null,
                    "datavalue__string": null,
                    "time": "+2013-12-03T00:00:00Z",
                    "timezone": 0,
                    "before": 0,
                    "after": 0,
                    "precision": 11,
                    "calendarmodel": "http://www.wikidata.org/entity/Q1985727",
                    "text": null,
                    "language": null
                  }
                }
              ]
            }
          ]
        }
      ],
      "P1015": [
        {
          "mainsnak": {
            "datavalue": {
              "id": null,
              "labels": null,
              "datavalue__string": "90196888"
            }
          },
          "qualifiers": null,
          "references": null
        }
      ],
      "P734": [
        {
          "mainsnak": {
            "datavalue": {
              "id": "Q351735",
              "labels": {
                "ar": "آدامز"
              },
              "datavalue__string": null
            }
          },
          "qualifiers": null,
          "references": [
            {
              "P248": [
                {
                  "datavalue": {
                    "id": "Q36578",
                    "labels": {
                      "de": "Gemeinsame Normdatei"
                    },
                    "datavalue__string": null,
                    "time": null,
                    "timezone": null,
                    "before": null,
                    "after": null,
                    "precision": null,
                    "calendarmodel": null,
                    "text": null,
                    "language": null
                  }
                }
              ],
              "P227": [
                {
                  "datavalue": {
                    "id": null,
                    "labels": null,
                    "datavalue__string": "119033364",
                    "time": null,
                    "timezone": null,
                    "before": null,
                    "after": null,
                    "precision": null,
                    "calendarmodel": null,
                    "text": null,
                    "language": null
                  }
                }
              ],
              "P407": [
                {
                  "datavalue": {
                    "id": "Q188",
                    "labels": {
                      "tg": "забони олмонӣ"
                    },
                    "datavalue__string": null,
                    "time": null,
                    "timezone": null,
                    "before": null,
                    "after": null,
                    "precision": null,
                    "calendarmodel": null,
                    "text": null,
                    "language": null
                  }
                }
              ],
              "P813": [
                {
                  "datavalue": {
                    "id": null,
                    "labels": null,
                    "datavalue__string": null,
                    "time": "+2022-10-09T00:00:00Z",
                    "timezone": 0,
                    "before": 0,
                    "after": 0,
                    "precision": 11,
                    "calendarmodel": "http://www.wikidata.org/entity/Q1985727",
                    "text": null,
                    "language": null
                  }
                }
              ]
            }
          ]
        }
      ],
      "P551": [
        {
          "mainsnak": {
            "datavalue": {
              "id": "Q159288",
              "labels": {
                "zh-hans": "圣巴巴拉"
              },
              "datavalue__string": null
            }
          },
          "qualifiers": {
            "P582": [
              {
                "datavalue": {
                  "precision": {
                    "precision__integer": 11,
                    "precision__number": null
                  },
                  "time": "+2001-05-11T00:00:00Z",
                  "timezone": 0,
                  "before": 0,
                  "after": 0,
                  "calendarmodel": "http://www.wikidata.org/entity/Q1985727",
                  "latitude": null,
                  "longitude": null,
                  "altitude": null,
                  "globe": null
                }
              }
            ]
          },
          "references": [
            {
              "P143": [
                {
                  "datavalue": {
                    "id": "Q328",
                    "labels": {
                      "en": "English Wikipedia"
                    },
                    "datavalue__string": null,
                    "time": null,
                    "timezone": null,
                    "before": null,
                    "after": null,
                    "precision": null,
                    "calendarmodel": null,
                    "text": null,
                    "language": null
                  }
                }
              ]
            }
          ]
        },
        {
          "mainsnak": {
            "datavalue": {
              "id": "Q84",
              "labels": {
                "en": "London"
              },
              "datavalue__string": null
            }
          },
          "qualifiers": {
            "P580": [
              {
                "datavalue": {
                  "precision": {
                    "precision__integer": 10,
                    "precision__number": null
                  },
                  "time": "+1952-09-00T00:00:00Z",
                  "timezone": 0,
                  "before": 0,
                  "after": 0,
                  "calendarmodel": "http://www.wikidata.org/entity/Q1985727",
                  "latitude": null,
                  "longitude": null,
                  "altitude": null,
                  "globe": null
                }
              }
            ]
          },
          "references": [
            {
              "P143": [
                {
                  "datavalue": {
                    "id": "Q328",
                    "labels": {
                      "en": "English Wikipedia"
                    },
                    "datavalue__string": null,
                    "time": null,
                    "timezone": null,
                    "before": null,
                    "after": null,
                    "precision": null,
                    "calendarmodel": null,
                    "text": null,
                    "language": null
                  }
                }
              ]
            }
          ]
        }
      ]
    }
  }
]
